docs/learn.md

```diff
@@ -721,20 +723,1250 @@ template](https://github.com/melange-re/melange-opam-template).
+variant and use the `bs.unwrap` attribute in the external.
+
+Let’s say we want to bind to the following JavaScript function:
+
+```js
+function padLeft(value, padding) {
+  if (typeof padding === "number") {
+    return Array(padding + 1).join(" ") + value;
+  }
+  if (typeof padding === "string") {
+    return padding + value;
+  }
+  throw new Error(`Expected string or number, got '${padding}'.`);
+}
+```
+
+Here, `padding` can be modeled as a polymorphic variant.
+
+```ocaml
+external padLeft:
+  string
+  -> ([ `Str of string
+      | `Int of int
+      ] [@bs.unwrap])
+  -> string
+  = "padLeft" [@@bs.val]
+
+let _ = padLeft "Hello World" (`Int 4)
+let _ = padLeft "Hello World" (`Str "Message from Melange: ")
+```
+
+Which produces the following JavaScript:
+
+```js
+padLeft("Hello World", 4);
+padLeft("Hello World", "Message from Melange: ");
+```
+
+As we saw in the [Non-shared data types](#non-shared-data-types) section, we
+should rather avoid passing variants directly to the JavaScript side. By using
+`bs.unwrap` we get the best of both worlds: from Melange we can use variants,
+while JavaScript gets the raw values inside them.
+
+#### Using polymorphic variants to bind to enums and string types
+
+Some JavaScript APIs take a limited subset of values as input. For example,
+Node’s `fs.readFileSync` second argument. It can only take a few given string
+values: `"ascii"`, `"utf8"`, etc.
+
+One could still bind it just as a string, but this would not prevent from using
+unsupported values. Let’s see how we can use polymorphic variants and the
+`bs.string` attribute to avoid runtime errors:
+
+```ocaml
+external read_file_sync :
+  name:string -> ([ `utf8 | `ascii ][@bs.string]) -> string = "readFileSync"
+  [@@bs.module "fs"]
+
+let _ = read_file_sync ~name:"xx.txt" `ascii
+```
+
+Which generates:
+
+```js
+var Fs = require("fs");
+Fs.readFileSync("xx.txt", "ascii");
 ```
-- intro: explain attributes, extensions, and `bs.` backwards compatibility
-- `melange.ppx` preprocessor and extensions
-- data types
-- call javascript from melange:
-  - intro to external (link to OCaml docs)
-  - bindings with external (to functions, objects or other values)
-  - global values
-- call melange from javascript
+
+Attaching `bs.string` to the polymorphic variant type makes its constructor
+compile to a string of the same name. This technique can be combined with the
+`bs.as` attributes to modify the strings produced from the polymorphic variant
+values, we will see an example of this just below.
+
+Aside from producing string values, Melange also offers `bs.int` to produce
+integer values:
+
+```ocaml
+external test_int_type :
+  ([ `on_closed | `on_open [@bs.as 20] | `in_bin ][@bs.int]) -> int
+  = "testIntType"
+  [@@bs.val]
+
+let value = test_int_type `on_open
+```
+
+In this example, `on_closed` will be encoded as 0, `on_open` will be 20 due to
+the attribute `bs.as` and `in_bin` will be 21.
```

Comment:
we need to explain why: because it starts counting up from the previous value.

Reply:
Added in https://github.com/melange-re/documentation-site/pull/12/commits/15499fb1f3e577830b4fc6e3a4e8c29d28d51b15.